### ISSUE TITLE
S3 Gateway with caching issues 503 instead of 404 for missing object in new bucket in response to head-object

### ISSUE THREAD

**Issue by rldleblanc** [minio/minio#7935]
With a S3 gateway and caching configured requesting a head-object on a non-existent object in a new bucket incorrectly returns a 503 error instead of a 404 error. Some fairly old buckets work okay and this seems to mostly affect newer buckets. Creating the bucket through AWS directly or through Minio have the same issue. Issuing the head-object directly to AWS properly returns the 404 error, but the same object through Minio returns a 503. Restarting Minio does not help.

```root@sun-gpu303:~/rleblanc# aws s3api head-object --endpoint-url http://minio-test --bucket rleblanc-s3-benchmark --key junk

An error occurred (503) when calling the HeadObject operation (reached max retries: 4): Service Unavailable
root@sun-gpu303:~/rleblanc# aws s3api head-object --bucket rleblanc-s3-benchmark --key junk

An error occurred (404) when calling the HeadObject operation: Not Found```

## Steps to Reproduce (for bugs)
1. Set up Minio S3 gateway with caching.
2. Create new bucket
3. Request head-object of non-existent object in new bucket
4. Observe 503 error instead of 404.

## Context
This was encountered by an s3-benchmark program to test the load of a new system. Practical application impacts are if an application wants to check if an object already exists and is expecting a 404 error.

## Regression
Unknown

## Your Environment
VERSION:
  2019-07-10T00:34:56Z
Ubuntu 16.04 linux binary
Linux sun-minio02 4.4.0-87-generic #110-Ubuntu SMP Tue Jul 18 12:55:35 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**Comment by nitisht** [comment]
Thanks for reporting @rldleblanc we'll take a look

**Comment by poornas** [comment]
@rldleblanc,  Can you share how you ran the [benchmark](https://github.com/dvassallo/s3-benchmark)? I ran mine like this `./s3-benchmark --endpoint http://localhost:9000` with no additional flags set, and the benchmark ran successfully against minio s3 gateway with caching enabled. Testing head object with aws-cli also gives a 404 as expected
```
(base) ➜  linux-amd64 git:(master) aws s3api head-object --endpoint-url http://localhost:9000 --bucket tbucket11 --key zzzt        

An error occurred (404) when calling the HeadObject operation: Not Found
(base) ➜  linux-amd64 git:(master) 
```
It might help debug this further if you can start http tracing with mc like this to see what might be going on? Appreciate if you can paste the log associated with this head-object operation including any retries,(after removing any sensitive info)
```
(base) ➜  linux-amd64 git:(master) mc admin trace mys3-alias --v
```

**Comment by rldleblanc** [comment]
The S3-benchmark is from https://github.com/dvassallo/s3-benchmark
`# ./s3-benchmark -bucket-name rleblanc-s3-test1 -endpoint http://minio-gw --full`
```
$ mc admin trace s3-gw01
18:22:55.17329 10.65.6.196:53192 s3.PutBucket minio-gw/rleblanc-s3-test1        200 OK
18:22:55.89562 10.65.6.196:53192 s3.HeadObject minio-gw/rleblanc-s3-test1/9b1454785f7363178e3de5fc2249abbab49013da      503 Service Unavailable
18:22:56.86953 10.65.6.196:53192 s3.HeadObject minio-gw/rleblanc-s3-test1/9b1454785f7363178e3de5fc2249abbab49013da      503 Service Unavailable
18:22:57.97102 10.65.6.196:53192 s3.HeadObject minio-gw/rleblanc-s3-test1/9b1454785f7363178e3de5fc2249abbab49013da      503 Service Unavailable
18:23:00.78663 10.65.6.196:53192 s3.HeadObject minio-gw/rleblanc-s3-test1/9b1454785f7363178e3de5fc2249abbab49013da      503 Service Unavailable
```

**Comment by poornas** [comment]
@rldleblanc, just to isolate this issue, what happens if caching is turned off at the gateway? unset MINIO_CACHE_DRIVES, and any other cache related env you might have set and then run the benchmark.

**Comment by poornas** [comment]
also , would help to see trace with the --verbose flag `mc admin trace --verbose mys3-alias`

**Comment by rldleblanc** [comment]
Disabled the cache.
```
minio-gw [REQUEST s3.PutBucket] [19:18:05.35122]
minio-gw PUT /rleblanc-s3-test1
minio-gw Host: minio-gw
minio-gw Accept-Encoding: gzip
minio-gw Authorization: AWS4-HMAC-SHA256 Credential=<KEY>/us-west-2/s3/aws4_request, SignedHeaders=content-length;host;x-amz-content-sha256;x-amz-date, Signature=d9ec1b72a5eff83679a491627c85df5872a1fb73c241cc88da91be2566120ceb
minio-gw Content-Length: 153
minio-gw User-Agent: aws-sdk-go/0.7.0 (go1.11.4; linux; amd64)
minio-gw X-Amz-Content-Sha256: a4a3d263179bfcbaf88755b14ea9db3048e4179c03bbc94ccf0f13c8c2d3df6a
minio-gw X-Amz-Date: 20190716T191805Z
minio-gw
minio-gw [RESPONSE] [19:18:05.87420]
minio-gw 409 Conflict
minio-gw Vary: Origin
minio-gw X-Amz-Request-Id: 15B1F9541E0E5EAC
minio-gw X-Xss-Protection: 1; mode=block
minio-gw Accept-Ranges: bytes
minio-gw Content-Length: 328
minio-gw Content-Security-Policy: block-all-mixed-content
minio-gw Content-Type: application/xml
minio-gw Server: MinIO/RELEASE.2019-07-10T00-34-56Z
minio-gw <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>BucketAlreadyOwnedByYou</Code><Message>Your previous request to create the named bucket succeeded and you already own it.</Message><BucketName>rleblanc-s3-test1</BucketName><Resource>/rleblanc-s3-test1</Resource><RequestId>15B1F9541E0E5EAC</RequestId><HostId></HostId></Error>                           
minio-gw
minio-gw [REQUEST s3.HeadObject] [19:18:05.87528]
minio-gw HEAD /rleblanc-s3-test1/9b1454785f7363178e3de5fc2249abbab49013da
minio-gw Host: minio-gw
minio-gw Authorization: AWS4-HMAC-SHA256 Credential=<KEY>/us-west-2/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=347ba414caa5e2c1b9c70505ef7511f9e48841c65f611742bff41ce49c549f95
minio-gw Content-Length: 0
minio-gw User-Agent: aws-sdk-go/0.7.0 (go1.11.4; linux; amd64)
minio-gw X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
minio-gw X-Amz-Date: 20190716T191805Z
minio-gw
minio-gw [RESPONSE] [19:18:05.89537]
minio-gw 503 Service Unavailable
minio-gw Content-Length: 0
minio-gw Content-Security-Policy: block-all-mixed-content
minio-gw Server: MinIO/RELEASE.2019-07-10T00-34-56Z
minio-gw Vary: Origin
minio-gw X-Amz-Request-Id: 15B1F9543D4C47F4
minio-gw X-Xss-Protection: 1; mode=block
minio-gw Accept-Ranges: bytes
minio-gw
minio-gw
minio-gw [REQUEST s3.HeadObject] [19:18:06.71351]
minio-gw HEAD /rleblanc-s3-test1/9b1454785f7363178e3de5fc2249abbab49013da
minio-gw Host: minio-gw
minio-gw Authorization: AWS4-HMAC-SHA256 Credential=<KEY>/us-west-2/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=34b6f58c3ec23560fe45467f53cf8b4f6285a0f0b2bb772150e866cdc00503e1
minio-gw Content-Length: 0
minio-gw User-Agent: aws-sdk-go/0.7.0 (go1.11.4; linux; amd64)
minio-gw X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
minio-gw X-Amz-Date: 20190716T191806Z
minio-gw
minio-gw [RESPONSE] [19:18:06.71865]
minio-gw 503 Service Unavailable
minio-gw Vary: Origin
minio-gw X-Amz-Request-Id: 15B1F9546F42B637
minio-gw X-Xss-Protection: 1; mode=block
minio-gw Accept-Ranges: bytes
minio-gw Content-Length: 0
minio-gw Content-Security-Policy: block-all-mixed-content
minio-gw Server: MinIO/RELEASE.2019-07-10T00-34-56Z
minio-gw
minio-gw
minio-gw [REQUEST s3.HeadObject] [19:18:08.64204]
minio-gw HEAD /rleblanc-s3-test1/9b1454785f7363178e3de5fc2249abbab49013da
minio-gw Host: minio-gw
minio-gw X-Amz-Date: 20190716T191808Z
minio-gw Authorization: AWS4-HMAC-SHA256 Credential=<KEY>/us-west-2/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=56666d02ce620f829f0f1534dc958063b02c75914081a5b9cf1400cba9f30340
minio-gw Content-Length: 0
minio-gw User-Agent: aws-sdk-go/0.7.0 (go1.11.4; linux; amd64)
minio-gw X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
minio-gw
minio-gw [RESPONSE] [19:18:08.64618]
minio-gw 503 Service Unavailable
minio-gw X-Amz-Request-Id: 15B1F954E235077E
minio-gw X-Xss-Protection: 1; mode=block
minio-gw Accept-Ranges: bytes
minio-gw Content-Length: 0
minio-gw Content-Security-Policy: block-all-mixed-content
minio-gw Server: MinIO/RELEASE.2019-07-10T00-34-56Z
minio-gw Vary: Origin
minio-gw
minio-gw
minio-gw [REQUEST s3.HeadObject] [19:18:12.63542]
minio-gw HEAD /rleblanc-s3-test1/9b1454785f7363178e3de5fc2249abbab49013da
minio-gw Host: minio-gw
minio-gw User-Agent: aws-sdk-go/0.7.0 (go1.11.4; linux; amd64)
minio-gw X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
minio-gw X-Amz-Date: 20190716T191812Z
minio-gw Authorization: AWS4-HMAC-SHA256 Credential=<KEY>/us-west-2/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=915f38085c109d62cfc444fbf53ca373f4842273a012b25b1e5f41f631830ec3
minio-gw Content-Length: 0
minio-gw
minio-gw [RESPONSE] [19:18:12.63924]
minio-gw 503 Service Unavailable
minio-gw X-Amz-Request-Id: 15B1F955D03B6C8D
minio-gw X-Xss-Protection: 1; mode=block
minio-gw Accept-Ranges: bytes
minio-gw Content-Length: 0
minio-gw Content-Security-Policy: block-all-mixed-content
minio-gw Server: MinIO/RELEASE.2019-07-10T00-34-56Z
minio-gw Vary: Origin
minio-gw
minio-gw
```

**Comment by poornas** [comment]
is the actual backend here Amazon S3 or Ceph?

**Comment by rldleblanc** [comment]
The backend is Amazon S3, Ceph is a little farther down the road.

**Comment by poornas** [comment]
@rldleblanc , seems like you are hitting S3 service limits here -https://docs.aws.amazon.com/AmazonS3/latest/dev/optimizing-performance.html.  minio doesn't set 503's at all or reinterpret the 404 - this 503 is being returned by the backend,

**Comment by rldleblanc** [comment]
That's odd. A new bucket is created and then a HEAD is requested, that is two operation. The limits are per bucket, so our other buckets should not be counted towards it. I don't think we are hitting those limits. It may be something else, but AWS directly is giving me a 404.

**Comment by rldleblanc** [comment]
I'm happy to run a debugging build or anything that could get you some more information. The debug above doesn't seem to show what AWS is returning.

**Comment by poornas** [comment]
Can you download this [binary](https://play.min.io/bin/minio?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=Q3AM3UQ867SPQQA43P2F%2F20190716%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190716T220351Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=5542cbae0a417d9d219b96036cb173c137290d406170d5f80bd52df7aadd5021) and paste the error you see on the server side? I just added a debug to see the error from S3

**Comment by rldleblanc** [comment]
I think there may be a bug in that build. I'm getting an EOF instead of 503 or 404.
```
panic: Filed to head S3 object: RequestError: send request failed
caused by: Head http://minio-gw/rleblanc-s3-test2/9b1454785f7363178e3de5fc2249abbab49013da: EOF
```

**Comment by poornas** [comment]
it was a build for linux-amd64 - you should have just seen something like `minio-go: HEAD returned : 404 ; 404 Not Found  err= <nil>` which is the debug stmt I added.

  Can you share stack trace for the panic - this `panic: Filed to head S3 object: RequestError: send request failed` doesn't come from minio source code. How did you start the minio gateway? What's your client - can you try `mc stat mys3-alias/bucket/object --debug` using the mc tool

**Comment by poornas** [comment]
nvm, it was from dvassallo/s3-benchmark repo
https://github.com/dvassallo/s3-benchmark/commit/1197ba17c556846f1b788756a168e6ab1279961a

the code here in this repo is panicking if HEAD returns a `Not Found`

**Comment by rldleblanc** [comment]
I'm seeing the requests for the bucket creation, but that's all I get. The s3-benchmark is getting an EOF unexpectedly and is panicing. It's like the HTTP session is being closed early.

**Comment by poornas** [comment]
@rldleblanc ,what's the error from minio server though? Were you able to start the gateway with this binary? A mc admin trace --v should show the HEAD request received by minio server. This panic is from the s3-benchmark tool

**Comment by poornas** [comment]
btw, can you also check if the benchmark works for you if the s3.amazonaws.com endpoint is specified for your creds instead of minio gateway

**Comment by rldleblanc** [comment]
I was able to start with this binary. This is all I got in the log:
```
minio-gw [REQUEST s3.PutBucket] [22:21:59.00091]
minio-gw PUT /rleblanc-s3-test2
minio-gw Host: minio-gw
minio-gw Content-Length: 153
minio-gw User-Agent: aws-sdk-go/0.7.0 (go1.11.4; linux; amd64)
minio-gw X-Amz-Content-Sha256: a4a3d263179bfcbaf88755b14ea9db3048e4179c03bbc94ccf0f13c8c2d3df6a
minio-gw X-Amz-Date: 20190716T222158Z
minio-gw Accept-Encoding: gzip
minio-gw Authorization: AWS4-HMAC-SHA256 Credential=<KEY>/us-west-2/s3/aws4_request, SignedHeaders=content-length;host;x-amz-content-sha256;x-amz-date, Signature=2e7d9addfdc94382fbfaa818807f95b6227f3fed467495c30c9b25fb8f2bd756
minio-gw 
minio-gw [RESPONSE] [22:21:59.72085]
minio-gw 200 OK
minio-gw Content-Security-Policy: block-all-mixed-content
minio-gw Location: /rleblanc-s3-test2
minio-gw Server: MinIO/DEVELOPMENT.2019-07-16T22-00-30Z
minio-gw Vary: Origin
minio-gw X-Amz-Request-Id: 15B2035D16D3614B
minio-gw X-Xss-Protection: 1; mode=block
minio-gw Accept-Ranges: bytes
minio-gw Content-Length: 0
minio-gw 
minio-gw
```

My creds only specify the key and secret key. I'm using "--endpoint" for minio and nothing to default AWS.

**Comment by rldleblanc** [comment]
I think you have an error in the debug statement and it is causing some error that is closing the TCP socket giving the client an EOF.

**Comment by poornas** [comment]
> I think you have an error in the debug statement and it is causing some error that is closing the TCP socket giving the client an EOF.

giving you a new [binary](https://play.min.io/bin/minio?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=Q3AM3UQ867SPQQA43P2F%2F20190716%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190716T235511Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=8044df153d8856ec374e2e79a14c680ed84417a4734173674dd6bfd00c09c46d) 

Just noticed that your bucket is on us-west-2, the benchmark creates a bucket and right away starts executing requests - there is usually a lag from bucket creation to availability, perhaps this is what you are seeing with the 503. Perhaps retry the benchmark after 10-15 mins on the same bucket and it should run successfully.

**Comment by poornas** [comment]
Or try with a bucket on us-east-1

**Comment by rldleblanc** [comment]
Okay, this is what I'm seeing on stderr:
`Jul 16 17:03:04 sun-minio01 minio[28204]: minio-go: err= Head http://rleblanc-s3-test2.s3.dualstack.us-west-1.amazonaws.com/9b1454785f7363178e3de5fc2249abbab49013da: 301 response missing Location header  resp: <nil>'`

**Comment by poornas** [comment]
ok, you need the --region flag while running benchmark to be us-west-1 if that's where you are creating bucket. It seems to be using us-west-2 as default if you look at their code and it is causing a redirect

**Comment by rldleblanc** [comment]
Still same error.
```
root@sun-gpu303:~/rleblanc# ./s3-benchmark -bucket-name rleblanc-s3-test2 -endpoint http://minio-gw -region us-west-1 -full

--- SETUP --------------------------------------------------------------------------------------------------------------------

Uploading 1 KB objects
   2% |                                        |  [0s:0s]panic: Filed to head S3 object: ServiceUnavailable: Service Unavailable
        status code: 503, request id: 15B20910C9B03934, host id:

goroutine 1 [running]:
main.setup()
        /Users/daniel/go/src/github.com/dvassallo/s3-benchmark/main.go:273 +0x997
main.main()
        /Users/daniel/go/src/github.com/dvassallo/s3-benchmark/main.go:113 +0x44
```
I thought MinIO was creating a new S3 request and not using much from the original request.

**Comment by poornas** [comment]
@rldleblanc , can you retry benchmark  with region flag set on the same bucket rleblanc-s3-test2 you specified for previous test?  To rule out whether 503 is due to bucket unavailability. btw, what did the stderr show with the new binary I sent?

**Comment by rldleblanc** [comment]
The last log message with the 301 is from the new binary. The old binary produced a huge stack trace. I tried us-west1, us-east-1 and I guess us-west-2 is the default. I didn't specify a region when I created the bucket, and since the s3-benchmark created the bucket with it's default region, that would be us-west-2.

So, if i'm not supplying a region and the bucket doesn't exist before running s3-benchmark (rleblanc-s3-test2 was created by the s3-benchmark), it seems to me that all traffic would be destined for us-west-2. That would complete the test that you are asking, correct?

**Comment by poornas** [comment]
that's right. I am also asking you to use the same bucket in the benchmark if  the run fails with 503, perhaps because bucket availability right away is not guaranteed. The second time you run the benchmark it would show a 409 conflict because bucket already exists - but I expect that the HEAD should pass.

**Comment by rldleblanc** [comment]
I'm not deleting the bucket after every run. I only deleted the rleblanc-s3-test1 bucket when the first debug binary failed (I had been using the same bucket all morning). The rleblanc-s3-test2 has had multiple runs against it as well. Most of the tests are on the same bucket created minutes or hours earlier to the same (default) region.

Why is AWS complaining that the Location header is missing in the 301?

**Comment by poornas** [comment]
@rldleblanc , when you start the minio gateway  the endpoint url passed in --address flag is having the us-west-1 as part of the url. This is the region that AWS will expect the bucket to be in, hence the 301 if region mismatches. 
`aws s3api get-bucket-location --endpoint-url http://minio-gw --bucket rleblanc-s3-test2` will show the actual region of the bucket. All requests to that bucket need the correct region in the Authorization header.

**Comment by harshavardhana** [comment]
Closing this as fixed by #7694 please re-open if not.
